src/mock/crypto.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Mock cryptographic primitives.
+
+pub type Digest256 = [u8; 32];
+// Note: using the real thing here as it seems to be already fast enough for tests.
+pub use tiny_keccak::sha3_256 as hash;
+
+pub mod signing {
+    pub use ed25519_dalek::SIGNATURE_LENGTH;
+    use ed25519_dalek::{PUBLIC_KEY_LENGTH, SECRET_KEY_LENGTH};
+    use rand_crypto::{CryptoRng, Rng};
+    use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+    use std::{
+        cmp::Ordering,
+        collections::hash_map::DefaultHasher,
+        fmt::{self, Debug, Display, Formatter},
+        hash::{Hash, Hasher},
+    };
+
+    #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+    pub struct PublicKey([u8; PUBLIC_KEY_LENGTH]);
+
+    impl PublicKey {
+        pub fn to_bytes(&self) -> [u8; PUBLIC_KEY_LENGTH] {
+            self.0
+        }
+
+        pub fn verify(&self, msg: &[u8], signature: &Signature) -> Result<(), SignatureError> {
+            let mut hasher = DefaultHasher::new();
+            hasher.write(msg);
+            hasher.write(&self.0);
+            let hash = hasher.finish().to_le_bytes();
+
+            if signature.0[..8] == hash {
+                Ok(())
+            } else {
+                Err(SignatureError)
+            }
+        }
+    }
+
+    impl<'a> From<&'a SecretKey> for PublicKey {
+        fn from(sk: &'a SecretKey) -> Self {
+            Self(sk.0)
+        }
+    }
+
+    pub struct SecretKey([u8; SECRET_KEY_LENGTH]);
+
+    impl SecretKey {
+        pub fn generate<T: CryptoRng + Rng>(rng: &mut T) -> Self {
+            Self(rng.gen())
+        }
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct Signature([u8; SIGNATURE_LENGTH]);
+
+    impl Signature {
+        pub fn from_bytes(bytes: &[u8]) -> Result<Signature, SignatureError> {
+            if bytes.len() < SIGNATURE_LENGTH {
+                Err(SignatureError)
+            } else {
+                let mut signature = Self([0; SIGNATURE_LENGTH]);
+                signature.0.copy_from_slice(bytes);
+                Ok(signature)
+            }
+        }
+    }
+
+    impl PartialEq for Signature {
```

Comment:
is that not deriving all these?

Reply:
Unfortunately no, as the array in `Signature` is 64 elements long and `derive` of `Eq`, `Ord`, `Hash`, etc... works only up to 32 elements. This will get resolved when const generics get stabilized in rust, which should hopefully be soon.